core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CommonMesosSchedulerBackend.scala

```diff
@@ -0,0 +1,249 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.scheduler.cluster.mesos
+
+import java.io.File
+import java.util.{List => JList}
+import java.util.Collections
+import java.util.concurrent.locks.ReentrantLock
+
+import com.google.common.collect.HashBiMap
+
+import scala.collection.JavaConversions._
+import scala.collection.mutable.{HashMap => MutableHashMap, HashSet => MutableHashSet}
+
+import org.apache.mesos.{Scheduler => MScheduler}
+import org.apache.mesos._
+import org.apache.mesos.Protos.{TaskInfo => MesosTaskInfo, TaskState => MesosTaskState, _}
+
+import org.apache.spark.{Logging, SparkContext, SparkEnv, SparkException}
+import org.apache.spark.scheduler.{SchedulerBackend, TaskSchedulerImpl}
+import org.apache.spark.scheduler.cluster.CoarseGrainedSchedulerBackend
+import org.apache.spark.util.{Utils, AkkaUtils}
+
+/**
+ * Shared code between {@link FineGrainedMesosSchedulerBackend} and
+ * {@link CoarseGrainedMesosSchedulerBackend}.
+ */
+trait CommonMesosSchedulerBackend extends SchedulerBackend {
+
+  self: MScheduler with Logging =>
```

Comment:
Personally I avoid using self-types unless the dependency is mutual. Otherwise, inheritance is just as good and usually simpler to understand. Also, less coupling